subxt/src/utils.rs

```diff
@@ -96,3 +96,137 @@ unsafe impl<T> Sync for PhantomDataSendSync<T> {}
+
+/// Generic `scale_bits` over `bitvec`-like `BitOrder` and `BitFormat` types.
+pub mod bits {
+    use codec::{
+        Compact,
+        Input,
+    };
+    use scale_bits::{
+        scale::format::{
+            Format,
+            OrderFormat,
+            StoreFormat,
+        },
+        Bits,
+    };
+    use std::marker::PhantomData;
+
+    macro_rules! store {
+        ($ident: ident; $(($ty: ident, $wrapped: ty)),*) => {
+            /// Associates `bitvec::store::BitStore` trait with corresponding, type-erased `scale_bits::StoreFormat` enum.
+            ///
+            /// Used to decode bit sequences in runtime by providing `scale_bits::StoreFormat` using
+            /// `bitvec`-like type type parameters.
+            pub trait $ident {
+                /// Corresponding `scale_bits::StoreFormat` value.
+                const FORMAT: StoreFormat;
+                /// Number of bits that the backing store types holds.
+                const BITS: u32;
+            }
+
+            $(
+                impl $ident for $wrapped {
+                    const FORMAT: StoreFormat = StoreFormat::$ty;
+                    const BITS: u32 = <$wrapped>::BITS;
+                }
+            )*
+        };
+    }
+
+    macro_rules! order {
+        ($ident: ident; $($ty: ident),*) => {
+            /// Associates `bitvec::order::BitOrder` trait with corresponding, type-erased `scale_bits::OrderFormat` enum.
+            ///
+            /// Used to decode bit sequences in runtime by providing `scale_bits::OrderFormat` using
+            /// `bitvec`-like type type parameters.
+            pub trait $ident {
+                /// Corresponding `scale_bits::OrderFormat` value.
+                const FORMAT: OrderFormat;
+            }
+
+            $(
+                #[doc = concat!("Type-level value that corresponds to `scale_bits::OrderFormat::", stringify!($ty), "` at run-time")]
+                #[doc = concat!(" and `bitvec::order::BitOrder::", stringify!($ty), "` at the type level.")]
+                #[derive(Clone, Debug, PartialEq, Eq)]
+                pub enum $ty {}
+                impl $ident for $ty {
+                    const FORMAT: OrderFormat = OrderFormat::$ty;
+                }
+            )*
+        };
+    }
+
+    store!(BitStore; (U8, u8), (U16, u16), (U32, u32), (U64, u64));
+    order!(BitOrder; Lsb0, Msb0);
+
+    /// Constructs a run-time format parameters based on the corresponding type-level parameters.
+    pub fn bit_format<Store: BitStore, Order: BitOrder>() -> Format {
+        Format {
+            order: Order::FORMAT,
+            store: Store::FORMAT,
+        }
+    }
+
+    /// `scale_bits::Bits` generic over the bit store (`u8`/`u16`/`u32`/`u64`) and bit order (LSB, MSB) types.
+    #[derive(Debug, Clone, PartialEq, Eq)]
+    pub struct DecodedBits<Store: BitStore, Order: BitOrder>(
+        pub Bits,
+        pub PhantomData<Store>,
+        pub PhantomData<Order>,
```

Comment:
One thing I forgot to add a comment for; instead of those all being `pub`, how about having an `into_bits()` method or something to pull the `Bits` out, and otherwise hiding the internals?

Reply:
Hiding the internals means hiding the default constructor and not being able to extract the wrapped value using a match pattern. Not saying it's greatly useful but I thought we wanted to be more transparent about this being a simple wrapper around `scale_bits::Bits`. Maybe we should add the `into_bits` convenience method but not hide the internals?

Reply:
That's true, but when would somebody need to match on it? I guess I think the `PhantomData` markers are a bit ugly to have as part of a public interface is all; perhaps they could be made not `pub` so that you can match on the `bits` while hiding those things?

Reply:
Unfortunately, hiding any member hides the constructor, so you can't do it, even with skipping the rest like so `let DecodedBits(bits, ..) = ...`. Do you mean matching on `value.0`?

Do you want me to only keep the functions from https://github.com/paritytech/subxt/pull/718/commits/2772f360aaee7139c789c7ef7cdecc6550326a10 and hide the internals?

Reply:
Ah ok! I think personally I prefer the `into_bits()` and `to_bits()` approach then (my pov I guess is that I prefer to hide internals and limit the public API surface so that we are freer to make changes that don't break, plus `into_bits()` et al just feels consistent with various other parts of the subxt APIs) :)

Reply:
Makes sense! Done in https://github.com/paritytech/subxt/pull/718/commits/8e58032c9e50e5b4696f9998b5516fa7feea3eed